contracts/pcv/sentinel/PCVSentinel.sol

```diff
@@ -0,0 +1,173 @@
+     * @dev you *can* use this on-chain if you want, but this is made
+     * to be called off-chain to pass into protecManys and not hyper-efficient
+     */
+    function checkAll()
+        external 
+        view
+        override 
+        returns (address[] memory) 
+    {
+        uint256 numGuardsCanProtec;
+
+        for (uint256 i=0; i<guards.length(); i++) {
+            if(IGuard(guards.at(i)).check()) {
+                numGuardsCanProtec++;
+            }
+        }
+
+        address[] memory guardsWhoCanProtec = new address[](numGuardsCanProtec);
+
+        for (uint256 i=0; i<guards.length(); i++) {
+            if(IGuard(guards.at(i)).check()) {
+                numGuardsCanProtec--;
+                guardsWhoCanProtec[numGuardsCanProtec] = guards.at(i);
+            }
+        }
+
+        return guardsWhoCanProtec;
+    }
+
+    // ---------- Governor-or-Admin-Or-Guardian-Only State-Changing API ----------
+
+    /**
+     * @notice adds a guard contract to the PCV Sentinel
+     * @param guard the guard-contract to add
+     */
+    function knight(address guard) 
+        external 
+        override 
+        isGovernorOrGuardianOrAdmin
+    {
+        guards.add(guard);
+
+        // Inform the kingdom of this glorious news
+        emit GuardAdded(guard);
+    }
+
+    /**
+     * @notice removes a guard
+     * @param traitor the guard-contract to remove
+     */
+    function slay(address traitor) 
+        external 
+        override 
+        isGovernorOrGuardianOrAdmin
+    {
+        guards.remove(traitor);
+
+        // Inform the kingdom of this sudden and inevitable betrayal
+        emit GuardRemoved(traitor);
+    }
+
+    // ---------- Public State-Changing API ----------
+
+    /**
+     * @notice checks the guard and runs its protec actions if needed
+     * @param guard the guard to activate
+     */
+    function protec(address guard) 
+        external
+        override
+    {
+        require(guards.contains(guard), "Guard does not exist.");
+
+        bool needsProtec = IGuard(guard).check();
+
+        if (needsProtec) {
+            (address[] memory targets, bytes[] memory datas) = IGuard(guard).getProtecActions();
+
+            for(uint256 i=0; i<targets.length; i++) {
+                require(targets[i] != address(this), "Nyeh!");
+                targets[i].call(datas[i]); // Unopinionated about reverts
+            }
+
+            emit Protected(guard);
+        } else {
+            revert("No need to protec.");
+        }
+    }
+
+    /**
+     * @notice activate many guards
+     * @dev we use try-catch here in case any guards fail, this could be the case
+     * if something changes before the transaction hits the chain
+     * @param whichGuards the guards to try to activate
+     */
+    function protecMany(address[] calldata whichGuards)
+        public 
+        override
+    {
+        for(uint256 i=0; i<whichGuards.length; i++) {
+            try this.protec(whichGuards[i]) {
```

Comment:
Swallowing errors is a huge defi antipattern.

I think we are far more at risk of missing something here than being screwed because we failed loud. Need some pretty specific contrary evidence to change my mind

Reply:
Require flawless execution of proective actions is an antipattern in security.
I agree that the risk of not knowing if something has failed is large, but I think the risk of not preventing a hack is much larger.

We could add in a bool that lets the caller decide though.

Reply:
I agree with Joey here, we should not just ignore errors when we are trying to protec.

Reply:
@Joeysantoro @ElliotFriedman neither of you have addressed the fact that the only way to hear about errors is events here.

> If someone calls into this contract from another contract and it reverts, we won't see those. We'll see the events if it doesn't revert though.

Therefore the best way to fail loudly is to *not* revert.

I added in a bool that lets the caller decide if it should revert or not.

Reply:
(it's important to remember here that we won't be the only ones calling into this contract)

Reply:
That is a case for only allowing all or nothing tx's.

Given that ethereum is a dark forest and a highly adversarial environment, we should not allow our contracts to halfway execute a transaction that is supposed to protect many of them at the same time.

Reply:
> That is a case for only allowing all or nothing tx's.
> 
> Given that ethereum is a dark forest and a highly adversarial environment, we should not allow our contracts to halfway execute a transaction that is supposed to protect many of them at the same time.

Sorry I don't really know what you're saying here